pkg/addonmanager/controllers/agentdeploy/default_hook_sync.go

```diff
@@ -0,0 +1,73 @@
+package agentdeploy
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/constants"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	"open-cluster-management.io/addon-framework/pkg/basecontroller/factory"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+type defaultHookSyncer struct {
+	controller *addonDeployController
+	agentAddon agent.AgentAddon
+}
+
+func (s *defaultHookSyncer) sync(ctx context.Context,
+	syncCtx factory.SyncContext,
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (*addonapiv1alpha1.ManagedClusterAddOn, error) {
+	installMode := constants.InstallModeDefault
+	deployWorkNamespace := addon.Namespace
+	_, hookWork, err := s.controller.buildManifestWorks(ctx, s.agentAddon, installMode, deployWorkNamespace, cluster, addon)
+	if err != nil {
+		return addon, err
+	}
+
+	if hookWork == nil {
+		addonRemoveFinalizer(addon, constants.PreDeleteHookFinalizer)
+		return addon, nil
+	}
+
+	if addonAddFinalizer(addon, constants.PreDeleteHookFinalizer) {
```

Comment:
this is tricky are using updateStatus to update finalizer? we should update finalizer separately here.

Reply:
my addon update process is:
I only add/remove finalizer to/from the addon in each sync.
in the `updateAddon` after all syncs,  check the finalizer change firstly, if has changes, update the addon and return, do not patch the condition, otherwise patch the condition.